### PR TITLE
Pay delivery fees from the assets available during execution

### PR DESCRIPTION
**Pull request by franciscoaguirre**
Implements XCM RFC https://github.com/paritytech/xcm-format/pull/53

Related: #3434

The problem this PR aims to address is the lack of functionality to pay for delivery fees directly with assets available during execution.
The way it solves it is by hijacking the `BuyExecution` instruction. After it has paid for execution, the unspent assets are used to pay for the delivery fees.

We put all unspent assets after paying for execution fees into a new `delivery_fees` register.
We could estimate how much to put and return the rest to the holding register, but estimating how much is needed is very complicated -- see commit history for how/why.

Because we just put all the assets into this register, some programs would probably need to change.
Before this change, there's no issue with users writing:

```rust
let xcm = Xcm(vec![
  WithdrawAsset(vec![asset]),
  BuyExecution { fees: asset, .. },
  DepositAsset { assets: asset.into(), .. },
]);
```

Now, this wouldn't deposit anything, since we take everything for fees.

Don't worry about the xcm simulator stuff, it's just to test this makes sense. I'll do better tests as I advance on this.

This is still a work-in-progress, but I'd appreciate any feedback about the approach, optimization concerns, security concerns, anything.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5317059

**Comment by xlc** [comment]
I just don't like it need to estimate the worst case fee first and then calculate the fee one more time later to actually charge the fee

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5322105

**Comment by franciscoaguirre** [comment]
I agree that's not the best. I needed a way to know how much to take from the unspent assets used in BuyExecution.

We could just charge the worst case. That would let us only calculate it once, but it might be a lot of overpaying. I'd need to do some benchmarking to see how much more expensive that would be.

If people specified a specific amount for fees then it wouldn't be a problem either. However, if they specify `All` as the filter, we could be keeping all their assets simply for fee payment. I could change the filter for only being able to specify concrete assets but that would be a bigger change and probably worse UX for the majority of users that don't want to care about fees.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5324434

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5324584

**Comment by xlc** [comment]
can we just have a fee register?

**Comment by franciscoaguirre** [comment]
But we do in this case. I can just put everything in `unspent` in the fee register if that's better

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325401

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325551

**Comment by franciscoaguirre** [comment]
It's clearly much simpler if we don't try to estimate the fees beforehand. I added that if there are assets left in the fee register after execution, then they're trapped. Seems like a simple solution that can work.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325696

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325714

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325715

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325871

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5325854

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5328515

File: polkadot/xcm/xcm-executor/src/lib.rs [hunk 1]
```diff
@@ -929,8 +950,44 @@ impl<Config: config::Config> XcmExecutor<Config> {
...
+					// We want to put some amount of these `unspent` assets into the `delivery_fees`
+					// register. We could either put all of them or try to estimate how much to put.
+					//
+					// If we just put all the assets into this register, some programs would
+					// probably need to change. Before this change, there's no issue with users
+					// writing:
+					// - WithdrawAsset(vec![asset])
+					// - BuyExecution { fees: asset, .. }
+					// - DepositAsset { assets: asset.into(), .. }
+					// Now, this wouldn't deposit anything, since we take everything for fees.
```
[the remark this answers is not in the document]
Comment (franciscoaguirre) [reply polkadot/xcm/xcm-executor/src/lib.rs]: I'll instead add a new instruction `DepositFee` to make sure no existing programs are broken

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5333662

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5333816

**Comment by franciscoaguirre** [comment]
I agree we should have more discussions about this, I created an RFC for it and mentioned it in the polkadot forum: https://forum.polkadot.network/t/xcm-rfc-better-fee-handling/6547/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/parity-tech-update-for-march/7226/1

**Comment by franciscoaguirre** [comment]
Closing in favor of https://github.com/paritytech/polkadot-sdk/pull/5420

**Pull request closed without merging**